File: app/fizzbuzz/src/FizzBuzz/Domain/Model/FizzBuzzValue.hs
module FizzBuzz.Domain.Model.FizzBuzzValue (
  FizzBuzzValue(..),
  createFizzBuzzValue,
  getValue,
  getNumber
) where

import FizzBuzz.Domain.Type.FizzBuzzError

-- 値オブジェクト：数値と結果文字列をカプセル化
data FizzBuzzValue = FizzBuzzValue 
  { number :: Int
  , value :: String
  } deriving (Show, Eq)

-- スマートコンストラクタ
createFizzBuzzValue :: Int -> String -> Either FizzBuzzError FizzBuzzValue
createFizzBuzzValue num val
  | num <= 0 = Left (InvalidInput num "正の値のみ有効です")
  | otherwise = Right (FizzBuzzValue num val)

-- 安全なアクセサ
getValue :: FizzBuzzValue -> String
getValue = value

getNumber :: FizzBuzzValue -> Int
getNumber = number

File: app/fizz_buzz/lib/values/fizz_buzz_value.ex
# FizzBuzzの値オブジェクト
defmodule FizzBuzzValue do
  @enforce_keys [:number, :value]
  defstruct [:number, :value]

  def new(number, value) do
    %FizzBuzzValue{number: number, value: value}
  end

  def get_number(%FizzBuzzValue{number: number}), do: number
  def get_value(%FizzBuzzValue{value: value}), do: value
  def to_string(%FizzBuzzValue{value: value}), do: value
end